### PR TITLE
Add Doxygen documentation

### PR DESCRIPTION
**Pull request by ish222**
- Add doxygen documentation for all new type traits
- Add CMake build target to build HTML documentation

### CODE REVIEW

File: include/opencl_type_traits [hunk 1]
```diff
@@ -678,6 +886,16 @@ using replace_reference_address_space_t =
 
 // add_<address_space>_reference_address_space
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __generic.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:891]:
```suggestion
 * the address space of type reference T refers to is replaced with __generic. 
```

File: include/opencl_type_traits [hunk 2]
```diff
@@ -688,6 +906,16 @@ template <typename T>
 using add_generic_reference_address_space_t =
     typename add_generic_reference_address_space<T>::type;
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __global.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:911]:
```suggestion
 * the address space of type reference T refers to is replaced with  __global. 
```

File: include/opencl_type_traits [hunk 3]
```diff
@@ -698,6 +926,16 @@ template <typename T>
 using add_global_reference_address_space_t =
     typename add_global_reference_address_space<T>::type;
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __private.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:931]:
```suggestion
 * the address space of type reference T refers to is replaced with __private. 
```

File: include/opencl_type_traits [hunk 4]
```diff
@@ -708,6 +946,16 @@ template <typename T>
 using add_private_reference_address_space_t =
     typename add_private_reference_address_space<T>::type;
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __local.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:951]:
```suggestion
 * the address space of type reference T refers to is replaced with __local. 
```

File: include/opencl_type_traits [hunk 5]
```diff
@@ -718,6 +966,16 @@ template <typename T>
 using add_local_reference_address_space_t =
     typename add_local_reference_address_space<T>::type;
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __constant.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:971]:
```suggestion
 * tthe address space of type reference T refers to is replaced with  __constant. 
```

File: include/opencl_type_traits [hunk 6]
```diff
@@ -718,6 +966,16 @@ template <typename T>
...
     typename add_local_reference_address_space<T>::type;
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __constant. 
+ *
+ * If T is not a reference type, then 
+ * type is the same as the T provided.
+ *
+ * @tparam T - a pointer type who's address space is set to __constant.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:976]:
```suggestion
 * @tparam T - a reference type with address space it refers to being replaced to __constant.
```

File: include/opencl_type_traits [hunk 7]
```diff
@@ -708,6 +946,16 @@ template <typename T>
...
     typename add_private_reference_address_space<T>::type;
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __local. 
+ *
+ * If T is not a reference type, then 
+ * type is the same as the T provided.
+ *
+ * @tparam T - a pointer type who's address space is set to __local.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:956]:
```suggestion
 * @tparam T - a reference type with address space it refers to being replaced to __local.
```

File: include/opencl_type_traits [hunk 8]
```diff
@@ -698,6 +926,16 @@ template <typename T>
...
     typename add_global_reference_address_space<T>::type;
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __private. 
+ *
+ * If T is not a reference type, then 
+ * type is the same as the T provided.
+ *
+ * @tparam T - a pointer type who's address space is set to __private.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:936]:
```suggestion
 * @tparam T - a reference type with address space it refers to being replaced to __private.
```

File: include/opencl_type_traits [hunk 9]
```diff
@@ -688,6 +906,16 @@ template <typename T>
...
     typename add_generic_reference_address_space<T>::type;
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __global. 
+ *
+ * If T is not a reference type, then 
+ * type is the same as the T provided.
+ *
+ * @tparam T - a pointer type who's address space is set to __global.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:916]:
```suggestion
 * @tparam T - a reference type with address space it refers to being replaced to __global.
```

File: include/opencl_type_traits [hunk 10]
```diff
@@ -678,6 +886,16 @@ using replace_reference_address_space_t =
...
 // add_<address_space>_reference_address_space
 
+/**
+ * Provides the member typedef type which is the same as T, except that
+ * the address space of reference T is replaced with __generic. 
+ *
+ * If T is not a reference type, then 
+ * type is the same as the T provided.
+ *
+ * @tparam T - a pointer type who's address space is set to __generic.
```
Comment (AnastasiaStulova) [inline include/opencl_type_traits:896]:
```suggestion
 * @tparam T - a reference type with address space it refers to being replaced to __generic.
```

**Comment by AnastasiaStulova** [comment]
I think we should document all vector type traits too although if the behavior strictly follows C++ we can just refer to C++ standard and/or use similar wording to C++ ref pages.

**Comment by ish222** [comment]
I'll add the vector documentation in a separate PR.

**Review by AnastasiaStulova — APPROVED** [review]
LGTM! Thanks

**Pull request merged**